Update Apex Writing and Testing

hbc-bindu opened the pull request:
### What does this PR do?

-- Update the Writing article with auto-completion suggestions displaying documentation for items in the list (W-7142561). Rewrite the section.
-- Rewrite Testing.
-- Delete Quick Fix (added it to the Writing section as it is a better fit there  - widget to automatically declare the method) and Apex Language Server article.


### What issues does this PR fix or reference?
#<Insert GitHub Issue>, @<Insert GUS WI>@

### Functionality Before
<insert gif and/or summary>

### Functionality After
<insert gif and/or summary>

lcampos commented on docs/articles/apex/writing.md:
```diff
@@ -26,51 +37,58 @@ You can preview, view, or go to definitions of:
+Whenever you refresh SObject definitions, the representative Apex classes are deleted and regenerated. You can modify the SObjects either by updating the objects’ `.object-meta.xml` and `.field-meta.xml` files; Or by making changes declaratively in the default org. After modifying the SObjects, make sure to sync your local project and the default org.
 
-To preview a definition, hold down Ctrl (Windows or Linux) or Cmd (macOS) and hover over the item whose definition you want to see.
+When you launch the Salesforce CLI Integration extension (which is part of the Salesforce Extension Pack) for the first time, `SFDX: Refresh SObject Definitions` command is executed automatically if your project doesn’t contain any Apex classes. After you add or edit standard or custom objects or their fields, make sure to run this command.
```
This is currently not true. For sobject refresh to run automatically the user needs to enable the `salesforcedx-vscode-apex.enable-sobject-refresh-on-startup` setting.

brpowell commented on docs/articles/apex/testing.md:
```diff
@@ -2,27 +2,39 @@
+## Explore Apex Tests
 
-The Apex Tests sidebar provides several features. Here, you can see all your Apex tests at a glance. You can run one test method, the test methods in one class, or all your tests. You can view the results of your last test run. And you can jump from those results to the corresponding lines in your code. To access this sidebar, click the beaker icon (hover text: Test) in the view bar on the left s ...
+Apex Tests view in the Side Bar provides several features for working with your tests such as run one test method, test methods in one class, or all your tests. To access Apex Tests view, click the beaker icon ({% octicon beaker %}) in the Activity Bar on the left side of the VS Code editor. If you don’t see this icon, make sure that the project contains an `sfdx-project.json` file in its root di ...
```
I'm wondering if we should be calling the test sidebar the "Apex Tests" view since the test view displays both apex and lwc tests now. But maybe that's overthinking it. @smaddox-sf

brpowell replied:
Should we link the sfdx project json section under the first project doc?

smaddox-sf replied:
That's a good point @brpowell .  I think we should just call it 'Tests' when we refer to the sidebar since it is more than Apex

brpowell commented on docs/articles/apex/testing.md:
```diff
@@ -2,27 +2,39 @@
 ---
 
-You can run Apex tests from within a file or from the Apex Tests sidebar. The sidebar provides other useful features for working with your tests.
+Apex extension allows you run Apex tests to verify the functionality of your code. You can retrieve code coverage results for your Apex classes and triggers every time you run one or more tests. To do this, edit your workspace settings and set `salesforcedx-vscode-core.retrieve-test-code-coverage` to **true**.
```
Should we link the doc on setting workspace settings?

brpowell commented on docs/articles/apex/testing.md:
```diff
@@ -2,27 +2,39 @@
+  - For passing tests, the blue icons next to the classes and methods change to green icons.
+  - For failing test, the icons turn red.
+
+You can view the test details by hovering over the test class name.
```
This seems a bit ambiguous. What test details, and is this referring to the test class name in the test view or in the editor window?

brpowell commented on docs/articles/apex/testing.md:
```diff
@@ -2,27 +2,39 @@
+
+- Go to Test Class Definitions
+
+To jump to the definition of a test class, a test method that passed, or a method that you haven’t run yet, click its name. If you click the name of a failed test method, you jump to the assert statement that failed.
```
```suggestion
To jump to the definition of a test class, a test method that passed, or a method that you haven’t run yet, click its name. If you click the name of a failed test method, you jump to the line where the failure occurred.
```

It could be an assert statement or just any failure like an exception.

brpowell commented on docs/articles/apex/testing.md:
```diff
@@ -2,27 +2,39 @@
+- Clear Test Results
 
-After you run Apex tests, two new commands are available in the command palette: **SFDX: Re-Run Last Invoked Apex Test Class** and **SFDX: Re-Run Last Invoked Apex Test Method**.
+To clear the test results, click the refresh icon at the top of the Apex Tests view (hover text: Refresh Tests).
```
The refresh button is also used for loading new tests if you've added more methods or classes since the last time the view was populated.

hbc-bindu replied:
Addedd another bullet Refresh Apex Tests View

brpowell commented on docs/articles/apex/writing.md:
```diff
@@ -2,18 +2,29 @@
 
-To see code-completion suggestions, press Ctrl+space when you’re working in a `.cls` or `.trigger` file. To navigate between the suggestions, use the arrow keys. To auto-complete a suggestion from the list, press Enter. To change how suggestions are pre-selected, see [IntelliSense - Suggestion selection](https://code.visualstudio.com/docs/editor/intellisense).
-![Animation showing code completion of a System.debug() statement](../../images/apex_completion.gif)
+Apex extension provides context-sensitive suggestions when you’re working in an Apex class or trigger. As you type, the auto-completion list of members such as, methods or variables pop up. The auto-completion list also displays the documentation for the suggestions. Use these keys:
```
Do we need to start the paragraphs in each section of this doc with "Apex extension"? For example, just say "Context-sensitive suggestions can be provided when working in an Apex class or trigger" or for below, "Code snippets are available for scaffolding class and interface definitions, and a variety of statements such as loops, conditionals, etc."

hbc-bindu replied:
Updated so that all the sections don't repeat 'Apex extension ...'

brpowell commented on docs/articles/apex/writing.md:
```diff
@@ -2,18 +2,29 @@
+![Animation showing code completion of PropertyController](../../images/apex_completion_with_doc.gif)
 
-## View or Jump to Definitions
+If you want to change how suggestions are pre-selected, see [IntelliSense - Suggestion selection](https://code.visualstudio.com/docs/editor/intellisense).
```
```suggestion
If you want to change how suggestions are pre-selected, see [IntelliSense - Suggestion selection](https://code.visualstudio.com/docs/editor/intellisense#_customizing-intellisense).
```

Minor change but this will take folks directly to the section on how to customize intellisense.

brpowell commented on docs/articles/apex/writing.md:
```diff
@@ -26,51 +37,58 @@ You can preview, view, or go to definitions of:
   - Object definitions
 
-(See the [Enable Code Smartness for SObjects](#enable-code-smartness-for-sobjects) section of this topic for information on working with standard and custom objects.)
+When you refresh SObject definitions, the extension uses the default org to generate Apex classes. These classes represent the standard and custom objects that the current user has access to. **Do not** edit these representative Apex classes, which are stored in a hidden directory on your local workstation.
```
```suggestion
When you refresh SObject definitions, the extension uses the default org to generate Apex classes. These classes represent the standard and custom objects that the current user has access to. **Do not** edit these representative Apex classes, which are stored in the hidden `.sfdx` directory in your project.
```

brpowell replied:
I'm torn about saying "__do not__" edit these classes if the user knows what they're doing. I'd maybe say it isn't recommended and explain that they're intended to be read-only and if they're modified, it may impact the autocomplete experience and any changes made may be wiped when rerunning the refresh command.

brpowell commented on docs/articles/apex/writing.md:
```diff
@@ -26,51 +37,58 @@ You can preview, view, or go to definitions of:
+- To view the list of the symbols in the file, press Ctrl+Shift+O (Windows or Linux) or Cmd+Shift+O (macOS).
+- To go to one of the symbols, select it from the list.
+
+![Outline view, showing the symbols in an Apex class](../../images/apex_outline.png)
```
We should also point out that there's also an actual outline section in the Explorer View
![Screen Shot 2020-02-06 at 11 09 59 AM](https://user-images.githubusercontent.com/10779873/73970044-546cc280-48d1-11ea-9af8-7cc0ba3efb1b.png)

hbc-bindu replied:
Updated the section with this info.

hbc-bindu replied:
@brpowell, @smaddox-sf , @lcampos , @sfsholden, Addressed all the review comments in Apex Writing and Testing topics. Please do let me know if you have any further feedback for these 2 topics.
Also, I created this branch based on an outdated branch and not the develop branch (facepalm). So I'll delete this pull request and create a new one...

sfsholden commented on docs/articles/user-guide/development-models.md:
```diff
@@ -0,0 +1,180 @@
+
+Salesforce Extensions for VS Code supports Org Development and Package Development models to authorize, create and switch orgs in your project:
+
+- **Org Development Model**: Allows you work with orgs that don’t have source tracking, such as sandbox, Developer Edition (DE) org, Trailhead Playground, or even a production org to retrieve and deploy code directly.
```
```suggestion
- **Org Development Model**: Allows you to work with orgs that don’t have source tracking. This includes Sandbox, Developer Edition (DE), Trailhead Playground, or Production orgs to retrieve and deploy code directly.
```

sfsholden commented on docs/articles/user-guide/development-models.md:
```diff
@@ -0,0 +1,180 @@
+
+To create developing against non-source-tracked orgs:
+
+1. Open the VS Code editor and from the Command Palette, run **SFDX: Create Project with Manifest**.
```
From a new user's perspective, I would want these steps included for bringing up the Command Palette:
(press Ctrl+Shift+P on Windows or Linux, or Cmd+Shift+P on macOS)

sfsholden commented on docs/articles/user-guide/development-models.md:
```diff
@@ -0,0 +1,180 @@
+
+## Org Development Model
+
+Use the Org Development model work with orgs that don’t have source tracking, such as sandboxes, Developer Edition (DE) orgs, or Trailhead Playgrounds, in Visual Studio Code. With this development model, you must track changes manually and deploy sets of changes to sandboxes and then to your production org. See the [Org Development Model](https://trailhead.salesforce.com/content/learn/modules/org ...
```
```suggestion
Use the Org Development model to work with orgs that don’t have source tracking, such as Sandboxes, Developer Edition (DE) orgs, or Trailhead Playgrounds, in Visual Studio Code. With this development model, you must track changes manually and deploy sets of changes to sandboxes and then to your production org. See the [Org Development Model](https://trailhead.salesforce.com/content/learn/modules/org-development-model) Trailhead module.
```

sfsholden commented on docs/articles/user-guide/development-models.md:
```diff
@@ -0,0 +1,180 @@
+To start developing with this model:
+
+- Create a project.
+- Use the Org Picker to authorize an org you want develop against.
```
```suggestion
- Use the Org Picker to authorize an org you want to develop against.
```

sfsholden commented on docs/articles/user-guide/development-models.md:
```diff
@@ -0,0 +1,180 @@
+
+## Create Project with Manifest
+
+To create developing against non-source-tracked orgs:
```
```suggestion
To start developing against non-source-tracked orgs:
```

smaddox-sf commented on docs/articles/apex/testing.md:
```diff
@@ -2,27 +2,39 @@
 ---
 
-You can run Apex tests from within a file or from the Apex Tests sidebar. The sidebar provides other useful features for working with your tests.
+Apex extension allows you run Apex tests to verify the functionality of your code. You can retrieve code coverage results for your Apex classes and triggers every time you run one or more tests. To do this, edit your workspace settings and set `salesforcedx-vscode-core.retrieve-test-code-coverage` to **true**.
```
> Should we link the doc on setting workspace settings?

Yeah, I think that's a good idea to link to it @brpowell .  Especially for new users, this might not be familiar and will save them searching

hbc-bindu commented:
Closing this pull request because I created the local branch based on an outdated branch. The updated topics after addressing the review comments: https://github.com/forcedotcom/salesforcedx-vscode/pull/1967
